Sources/NIO/NonBlockingFileIO.swift

```diff
@@ -214,6 +214,54 @@ public struct NonBlockingFileIO {
         }
     }
 
+    /// Write `byteCount` bytes from `buffer` to `fileHandle in `NonBlockingFileIO`'s private thread pool which is separate from any `EventLoop` thread.
+    ///
+    /// The returned `ByteBuffer` will not have less than `byteCount` bytes unless we hit end-of-file in which
+    /// case the `ByteBuffer` will contain the bytes available to read.
+    ///
+    /// - parameters:
+    ///   - fileHandle: The `FileHandle` to write to.
+    ///   - buffer: The `ByteBuffer` to write.
+    ///   - byteCount: The number of bytes to write from `buffer` to `fileHandle`.
+    ///   - eventLoop: The `EventLoop` to create the returned `EventLoopFuture` from.
+    /// - returns: An `EventLoopFuture` which is fulfilled if the write was successful or fails on error.
+    public func write(fileHandle: FileHandle,
+                      buffer: ByteBuffer,
+                      byteCount: Int,
+                      eventLoop: EventLoop) -> EventLoopFuture<()> {
+        guard byteCount > 0 else {
+            return eventLoop.newSucceededFuture(result: ())
+        }
+
+        var buf = buffer
+        return self.threadPool.runIfActive(eventLoop: eventLoop) {
+            var bytesWritten = 0
+            while bytesWritten < byteCount {
+                let n = try buf.readWithUnsafeMutableReadableBytes { ptr in
+                    let res = try fileHandle.withUnsafeFileDescriptor { descriptor in
+                        try Posix.write(descriptor: descriptor,
+                                        pointer: ptr.baseAddress!,
+                                        size: byteCount - bytesWritten)
+                    }
+                    switch res {
+                    case .processed(let n):
+                        assert(n >= 0, "write claims to have read a negative number of bytes \(n)")
+                        return n
+                    case .wouldBlock:
+                        throw Error.descriptorSetToNonBlocking
+                    }
+                }
+
+                if n == 0 {
```

Comment:
This end condition doesn't seem necessary: we're looping to handle this case. I think we can avoid the conditional branch here and unconditionally add `n`. We can `precondition` if you want the check, but we should never really enter this branch.